devito/ir/clusters/algorithms.py

```diff
@@ -1,75 +1,182 @@
-from collections import OrderedDict, namedtuple
+from collections import OrderedDict
 
-from devito.ir.clusters.cluster import Cluster
+from devito.ir.support import Scope
+from devito.ir.clusters.cluster import PartialCluster, ClusterGroup
 from devito.ir.dfg import TemporariesGraph
-from devito.ir.support import Stencil
 from devito.symbolics import xreplace_indices
 from devito.types import Scalar
+from devito.tools import flatten
 
-__all__ = ['clusterize', 'optimize']
+__all__ = ['clusterize', 'groupby']
 
 
-def merge(clusters):
+def groupby(clusters):
     """
-    Given an ordered collection of :class:`Cluster` objects, return a
-    (potentially) smaller sequence in which clusters with identical stencil
-    have been merged into a single :class:`Cluster`.
+    Given an ordered collection of :class:`PartialCluster` objects, return a
+    (potentially) smaller sequence in which dependence-free PartialClusters with
+    identical stencil have been squashed into a single PartialCluster.
```

Comment:
:class:`PartialCluster` for pretty-docs.

Reply:
fixing in the PR based on top of this (#423 ), thanks